### PR TITLE
[UX - SkyServe] user now is able to select a LB policy from a range of options

### PR DESCRIPTION
**Pull request by AlexCuadron**
I added the option to specify different LB policies, by default, RoundRobin is used. This PR is intended to enable easy switching between LB policies to facilitate its development.

The default behaviour (without user interaction) doesn't modify the execution flow and the user is not allowed to use any other LB policy other than round-robing without them being added explicitly.

Tested (run the relevant ones):

- [X] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: sky/serve/load_balancer.py [hunk 1]
```diff
@@ -27,18 +27,20 @@ class SkyServeLoadBalancer:
     policy.
     """
 
-    def __init__(self, controller_url: str, load_balancer_port: int) -> None:
+    def __init__(
+            self, controller_url: str, load_balancer_port: int,
+            load_balancing_policy: lb_policies.LoadBalancingPolicy) -> None:
```
Comment (cblmemo) [inline sky/serve/load_balancer.py:32]: Lets pass in the lb policy name here?

File: sky/serve/load_balancer.py [hunk 2]
```diff
@@ -223,9 +225,15 @@ async def startup():
         uvicorn.run(self._app, host='0.0.0.0', port=self._load_balancer_port)
 
 
-def run_load_balancer(controller_addr: str, load_balancer_port: int):
+def run_load_balancer(controller_addr: str, load_balancer_port: int,
+                      policy_name: str):
+    if policy_name == 'round_robin':
+        policy = lb_policies.RoundRobinPolicy()
+        raise ValueError(f'Unknown load balancing policy: {policy_name}')
```
Comment (cblmemo) [inline sky/serve/load_balancer.py:232]: For this, we might want to refer to the recovery strategy implementation:

1. Specify a registry from name to the class (RECOVERY_STRATEGIES in this example);
2. Define an function in the base class to choose from this dictionary.

https://github.com/skypilot-org/skypilot/blob/292febc85cee658bd4370dd1894f2a6c97a5e264/sky/jobs/recovery_strategy.py#L85-L112

This would help up to keep track of all lb policy in one places ;)

**Review by cblmemo — COMMENTED** [review]
This is awesome @AlexCuadron ! It will greatly expand our customizability. Left some discussion ;) We might also think of how to expose this feature to our end user, in our [Service YAML](https://skypilot.readthedocs.io/en/latest/serving/service-yaml-spec.html) - maybe add a `load_balancing_policy` section under `service`?

**Comment by AlexCuadron** [comment]
@cblmemo  Done! PTAL again :D

File: examples/serve/gorilla/gorilla.yaml
Comment (cblmemo) [inline examples/serve/gorilla/gorilla.yaml:1]: Lets set a default value for this ;) so we dont need to modify all those example files

File: sky/serve/load_balancer.py [hunk 3]
```diff
@@ -27,21 +27,27 @@ class SkyServeLoadBalancer:
...
         self._load_balancer_port: int = load_balancer_port
-        self._load_balancing_policy: lb_policies.LoadBalancingPolicy = (
-            lb_policies.RoundRobinPolicy())
+        # Use the registry to create the load balancing policy
+        self._load_balancing_policy = lb_policies.LoadBalancingPolicy.make(
+            load_balancing_policy_name)
         self._request_aggregator: serve_utils.RequestsAggregator = (
             serve_utils.RequestTimestamp())
-        # TODO(tian): httpx.Client has a resource limit of 100 max connections
+        # TODO: httpx.Client has a resource limit of 100 max connections
```
Comment (cblmemo) [inline sky/serve/load_balancer.py:50]: Why change this?

File: sky/serve/load_balancer.py [hunk 4]
```diff
@@ -223,10 +229,25 @@ async def startup():
         uvicorn.run(self._app, host='0.0.0.0', port=self._load_balancer_port)
 
 
-def run_load_balancer(controller_addr: str, load_balancer_port: int):
-    load_balancer = SkyServeLoadBalancer(controller_url=controller_addr,
-                                         load_balancer_port=load_balancer_port)
-    load_balancer.run()
+def run_load_balancer(controller_addr: str,
+                      load_balancer_port: int,
+                      policy_name: Optional[str] = None):
```
Comment (cblmemo) [inline sky/serve/load_balancer.py:234]:
```suggestion
                      load_balancing_policy_name: Optional[str] = None):
```
Lets align the argument name

File: sky/serve/load_balancer.py [hunk 5]
```diff
@@ -223,10 +229,25 @@ async def startup():
...
+            None.
+    """
+    try:
+        load_balancer = SkyServeLoadBalancer(
+            controller_url=controller_addr,
+            load_balancer_port=load_balancer_port,
+            load_balancing_policy_name=policy_name)
+        load_balancer.run()
+    except ValueError as e:
+        logger.error(f'Failed to start load balancer: {e}')
```
Comment (cblmemo) [inline sky/serve/load_balancer.py:250]: we probably dont want to silently allow the failure to happen. Remove the try-except block?

File: sky/serve/load_balancing_policies.py [hunk 6]
```diff
@@ -10,6 +10,10 @@
 
 logger = sky_logging.init_logger(__name__)
 
+# Define a registry for load balancing policies
+LB_POLICIES = {}
+DEFAULT_LB_POLICY = 'round_robin'
```
Comment (cblmemo) [inline sky/serve/load_balancing_policies.py:15]:
```suggestion
DEFAULT_LB_POLICY = None
```
should this be none?

File: sky/serve/load_balancing_policies.py [hunk 7]
```diff
@@ -25,6 +29,25 @@ class LoadBalancingPolicy:
...
+    @classmethod
+    def make(cls, policy_name: Optional[str] = None) -> 'LoadBalancingPolicy':
+        """Create a load balancing policy from a name."""
+        if policy_name is None:
+            policy_name = DEFAULT_LB_POLICY
+            return LB_POLICIES[policy_name]()
+
+        if policy_name not in LB_POLICIES:
+            raise ValueError(f'Unknown load balancing policy: {policy_name}')
+        return LB_POLICIES[policy_name]()
```
Comment (cblmemo) [inline sky/serve/load_balancing_policies.py:49]:
```suggestion
    @classmethod
    def make(cls, policy_name: Optional[str] = None) -> 'LoadBalancingPolicy':
        """Create a load balancing policy from a name."""
        if policy_name is None:
            policy_name = DEFAULT_LB_POLICY
        if policy_name not in LB_POLICIES:
            raise ValueError(f'Unknown load balancing policy: {policy_name}')
        return LB_POLICIES[policy_name]()
```

File: sky/serve/service_spec.py [hunk 8]
```diff
@@ -233,6 +237,8 @@ def add_if_not_none(section, key, value, no_empty: bool = False):
                         self.upscale_delay_seconds)
         add_if_not_none('replica_policy', 'downscale_delay_seconds',
                         self.downscale_delay_seconds)
+        add_if_not_none('service', 'load_balancing_policy',
+                        self._load_balancing_policy)
```
Comment (cblmemo) [inline sky/serve/service_spec.py:241]:
```suggestion
        add_if_not_none('load_balancing_policy', None
                        self._load_balancing_policy)
```
should this be none? you can try to debug with from_yaml_config(service.to_yaml_config()) and checked if everying is the same.

**Review by cblmemo — COMMENTED** [review]
Thanks @AlexCuadron for the awesome work! It mostly looks good to me. Left some discussions ;)

File: sky/serve/load_balancing_policies.py (hunk 6 above)
Comment (AlexCuadron) [reply sky/serve/load_balancing_policies.py:15]: If we define this as `None` this code section is not useful:
```
        if policy_name is None:
            policy_name = DEFAULT_LB_POLICY
            return LB_POLICIES[policy_name]()
```

File: sky/serve/service_spec.py (hunk 8 above)
Comment (AlexCuadron) [reply sky/serve/service_spec.py:241]: True, good catch! 👍

File: sky/serve/load_balancing_policies.py (hunk 7 above)
Comment (AlexCuadron) [reply sky/serve/load_balancing_policies.py:49]: Oops

File: sky/serve/load_balancer.py (hunk 5 above)
Comment (AlexCuadron) [reply sky/serve/load_balancer.py:250]: agreed

**Comment by AlexCuadron** [comment]
Thanks for the comments @cblmemo, fixed and ready for next round 💪

File: examples/serve/gorilla/gorilla.yaml
Comment (AlexCuadron) [reply examples/serve/gorilla/gorilla.yaml:1]: I added the parameter for illustrative purposes (eg, the user learns it can select a load-balancing policy no matter which example he/she looks at), `round-robin` is the default.

File: sky/serve/load_balancer.py (hunk 3 above)
Comment (AlexCuadron) [reply sky/serve/load_balancer.py]: Oops

File: examples/serve/gorilla/gorilla.yaml
Comment (cblmemo) [reply examples/serve/gorilla/gorilla.yaml:1]: Lets add a single YAML for the illustrative purpose ;) Adding those to existing examples would make it a little bit messy.

File: sky/serve/load_balancing_policies.py [hunk 9]
```diff
@@ -25,6 +29,24 @@ class LoadBalancingPolicy:
...
+            DEFAULT_LB_POLICY = name
+
+    @classmethod
+    def make(cls, policy_name: Optional[str] = None) -> 'LoadBalancingPolicy':
+        """Create a load balancing policy from a name."""
+        if policy_name is None:
+            policy_name = DEFAULT_LB_POLICY
+
+        if policy_name not in LB_POLICIES:
+            raise ValueError(f'Unknown load balancing policy: {policy_name}')
```
Comment (cblmemo) [inline sky/serve/load_balancing_policies.py:47]: Lets also check this in `SkyServiceSpec`, before the user launch the service. If the user has a typo in the yaml, we might want to let the user konw it before launch a controller.

https://github.com/skypilot-org/skypilot/blob/c6ae536d8dfedc3bbcf427a81480382b9d5f4c29/sky/resources.py#L810-L823

**Review by cblmemo — COMMENTED** [review]
Thanks for the quick fix @AlexCuadron ! Mostly looks good to me. Left some discussions.

btw, I created a branch here and lets merge our PR to the branch first. Merging into master might need more time and we want to move fast ;)

https://github.com/skypilot-org/skypilot/tree/heterogeneous-lb

File: sky/serve/load_balancing_policies.py (hunk 9 above)
Comment (AlexCuadron) [reply sky/serve/load_balancing_policies.py:47]: Super good point 👍

**Comment by AlexCuadron** [comment]
I changed the base branch and updated based on comments :)
PTAL @cblmemo

File: examples/serve/misc/cancel/service.yaml
Comment (cblmemo) [inline examples/serve/misc/cancel/service.yaml:1]: Do we really need to update those examples? some of them are used by the smoke test and it might break them. We also keep the empty lines for separation of the YAML file sections. Can we invert those unnecessary changes?

File: examples/serve/llama2/llama2.yaml [hunk 10]
```diff
@@ -15,6 +15,7 @@
 service:
   readiness_probe: /v1/models
   replicas: 2
+  load_balancing_policy: round_robin
```
Comment (cblmemo) [inline examples/serve/llama2/llama2.yaml:18]: lets write a new example instead of updating existing one.

File: sky/serve/load_balancer.py [hunk 11]
```diff
@@ -241,5 +259,11 @@ def run_load_balancer(controller_addr: str, load_balancer_port: int):
                         required=True,
                         default=8890,
                         help='The port where the load balancer listens to.')
+    parser.add_argument(
+        '--load-balancing-policy',
+        choices=['round_robin'],
```
Comment (cblmemo) [inline sky/serve/load_balancer.py:264]: lets update this to the keys of the registry?

File: sky/serve/service_spec.py [hunk 12]
```diff
@@ -7,6 +7,7 @@
 import yaml
 
 from sky.serve import constants
+from sky.serve.load_balancing_policies import LB_POLICIES
```
Comment (cblmemo) [inline sky/serve/service_spec.py:10]: nit: import modules instead of functions/classes.
https://google.github.io/styleguide/pyguide.html#224-decision

File: sky/serve/service_spec.py [hunk 13]
```diff
@@ -77,6 +79,13 @@ def __init__(
                     'Currently, SkyServe will cleanup failed replicas'
                     'and auto restart it to keep the service running.')
 
+        # Add the check for unknown load balancing policies
+        if load_balancing_policy is not None and \
+            load_balancing_policy not in LB_POLICIES:
```
Comment (cblmemo) [inline sky/serve/service_spec.py:84]:
```suggestion
        if (load_balancing_policy is not None and
            load_balancing_policy not in LB_POLICIES):
```
nit: Do not use a backslash for explicit line continuation. Instead, make use of Python’s implicit line joining inside parentheses, brackets and braces.
https://google.github.io/styleguide/pyguide.html#32-line-length

File: sky/serve/service_spec.py (hunk 12 above)
Comment (cblmemo) [reply sky/serve/service_spec.py:10]: Also, lets expose this to `sky/serve/__init__.py`?

https://github.com/skypilot-org/skypilot/blob/63e96f49a94a3bdd0263baf52de5eb746b4adc77/sky/jobs/__init__.py#L14

**Review by cblmemo — COMMENTED** [review]
Thanks for the prompt fix @AlexCuadron ! Mostly looks good to me. Left some nits :))

File: sky/serve/load_balancer.py (hunk 11 above)
Comment (AlexCuadron) [reply sky/serve/load_balancer.py:264]: Good point! 💯

**Comment by AlexCuadron** [comment]
Done! PTAL @cblmemo  :D

File: examples/serve/load_balancing_policies_example.yaml
Comment (cblmemo) [inline examples/serve/load_balancing_policies_example.yaml:1]: nit: eol of last line

File: sky/serve/load_balancer.py [hunk 14]
```diff
@@ -241,5 +259,12 @@ def run_load_balancer(controller_addr: str, load_balancer_port: int):
                         required=True,
                         default=8890,
                         help='The port where the load balancer listens to.')
+    parser.add_argument(
+        '--load-balancing-policy',
+        choices=list(lb_policies.LB_POLICIES.keys()),
+        default='round_robin',
+        help=f'The load balancing policy to use. Available policies: '
+        f'{", ".join(lb_policies.LB_POLICIES.keys())}.')
```
Comment (cblmemo) [inline sky/serve/load_balancer.py:267]:
```suggestion
    available_policies = list(lb_policies.LB_POLICIES.keys())
    parser.add_argument(
        '--load-balancing-policy',
        choices=available_policies,
        default='round_robin',
        help=f'The load balancing policy to use. Available policies: '
        f'{", ".join(available_policies)}.')
```
nit

File: sky/serve/__init__.py [hunk 15]
```diff
@@ -1,6 +1,7 @@
 """Modules for SkyServe services."""
 import os
 
+from sky.serve import load_balancing_policies as lb_policies
```
Comment (cblmemo) [inline sky/serve/__init__.py:4]:
```suggestion
from sky.serve.load_balancing_policies import LB_POLICIES
```
lets directly expose this ;)

**Review by cblmemo — COMMENTED** [review]
Thanks for the prompt fix @AlexCuadron ! Left some final nits 🚀

**Comment by AlexCuadron** [comment]
Done! PTAL! @cblmemo

**Comment by AlexCuadron** [comment]
@cblmemo or @Michaelvll  What changes are needed for this PR to be merged into the master branch? That way, we won’t need to keep updating the #heterogeneous-lb branch.

File: sky/serve/service_spec.py (hunk 12 above)
Comment (cblmemo) [inline sky/serve/service_spec.py:10]:
```suggestion
from sky import serve
```

File: sky/serve/service_spec.py [hunk 16]
```diff
@@ -77,6 +79,13 @@ def __init__(
                     'Currently, SkyServe will cleanup failed replicas'
                     'and auto restart it to keep the service running.')
 
+        # Add the check for unknown load balancing policies
+        if (load_balancing_policy is not None and
+                load_balancing_policy not in LB_POLICIES):
```
Comment (cblmemo) [inline sky/serve/service_spec.py:84]:
```suggestion
                load_balancing_policy not in serve.LB_POLICIES):
```

File: sky/serve/__init__.py (hunk 15 above)
Comment (cblmemo) [inline sky/serve/__init__.py:4]:
```suggestion
from sky.serve.load_balancing_policies import LB_POLICIES
```

**Review by cblmemo — APPROVED** [review]
Actually, for exposing `LB_POLICIES`, pls check the comments. And I do think this should be able to merge to master - cc @Michaelvll for a look. @AlexCuadron I've changed the base branch to master, could you help merging the latest one?

**Review by cblmemo — CHANGES_REQUESTED** [review]
Sorry, doesn't mean to approve it - lets wait for other ppl's opinions.

**Comment by AlexCuadron** [comment]
I have fixed it according to comments @cblmemo. PTAL @Michaelvll  :)

**Comment by AlexCuadron** [comment]
PTAL @Michaelvll

File: sky/serve/load_balancer.py [hunk 17]
```diff
@@ -223,9 +229,21 @@ async def startup():
         uvicorn.run(self._app, host='0.0.0.0', port=self._load_balancer_port)
 
 
-def run_load_balancer(controller_addr: str, load_balancer_port: int):
-    load_balancer = SkyServeLoadBalancer(controller_url=controller_addr,
-                                         load_balancer_port=load_balancer_port)
+def run_load_balancer(controller_addr: str,
+                      load_balancer_port: int,
+                      load_balancing_policy_name: Optional[str] = None):
```
Comment (cblmemo) [inline sky/serve/load_balancer.py:234]:
```suggestion
                      load_balancing_policy_name: Optional[str] = None) -> None:
```
nit

**Review by cblmemo — COMMENTED** [review]
Hi @AlexCuadron ! Thanks for adding this. I tried this PR and got the following:

```bash
$ sky serve up examples/serve/load_balancing_policies_example.yaml
Service from YAML spec: examples/serve/load_balancing_policies_example.yaml
ValueError: Invalid service YAML: Found unsupported field 'load_balancing_policy'.
```

Should we also update the `sky/utils/schemas.py`?

File: sky/serve/service_spec.py [hunk 18]
```diff
@@ -55,6 +57,12 @@ def __init__(
                 raise ValueError('readiness_path must start with a slash (/). '
                                  f'Got: {readiness_path}')
 
+        # Add the check for unknown load balancing policies
+        if (load_balancing_policy is not None and
+                load_balancing_policy not in serve.LB_POLICIES):
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    f'Unknown load balancing policy: {load_balancing_policy}')
```
Comment (cblmemo) [inline sky/serve/service_spec.py:65]: Should we print out all available policies here?

File: sky/serve/service_spec.py
Comment (cblmemo) [inline sky/serve/service_spec.py:1]: Lets also update the `__repr__` field to show the LB policy name if it is not None ;) (e.g. user explicitly specified policy)

File: sky/utils/schemas.py [hunk 19]
```diff
@@ -382,6 +382,10 @@ def get_service_schema():
             'replicas': {
                 'type': 'integer',
             },
+            'load_balancing_policy': {
+                'type': 'string',
+                'enum': ['round_robin']
```
Comment (cblmemo) [inline sky/utils/schemas.py:387]:
```suggestion
```
This would require 2 place to change when we add policy. Maybe we could just remove it and check it in the code?

**Comment by AlexCuadron** [comment]
PTAL @cblmemo

File: sky/serve/service_spec.py [hunk 20]
```diff
@@ -250,12 +264,16 @@ def autoscaling_policy_str(self):
...
+        if self.load_balancing_policy is not None:
+            policy_name = self.load_balancing_policy
+            lb_policy_str = f'\nLoad balancing policy:           {policy_name}'
         return textwrap.dedent(f"""\
             Readiness probe method:           {self.probe_str()}
             Readiness initial delay seconds:  {self.initial_delay_seconds}
             Readiness probe timeout seconds:  {self.readiness_timeout_seconds}
             Replica autoscaling policy:       {self.autoscaling_policy_str()}
-            Spot Policy:                      {self.spot_policy_str()}
+            Spot Policy:                      {self.spot_policy_str()}{lb_policy_str}
```
Comment (cblmemo) [inline sky/serve/service_spec.py:276]:
```suggestion
            Spot Policy:                      {self.spot_policy_str()}
            Load Balancing Policy:            {self.load_balancing_policy}
```

Seems like this is causing some problems due to the textwrap.dedent. Maybe lets just show the policy name no matter if it is explicit or not

```bash
ssu examples/serve/load_balancing_policies_example.yaml
Service from YAML spec: examples/serve/load_balancing_policies_example.yaml
Service Spec:
            Readiness probe method:           GET /health
            Readiness initial delay seconds:  20
            Readiness probe timeout seconds:  15
            Replica autoscaling policy:       Autoscaling from 2 to 4 replicas (target QPS per replica: 1)
            Spot Policy:                      No spot fallback policy
Load balancing policy:           round_robin
```

**Review by cblmemo — COMMENTED** [review]
Thanks for the quick fix @AlexCuadron ! Left 2 final nits ;)

**Comment by AlexCuadron** [comment]
Oops, sorry for the circular import 😅 
Should be gtg @cblmemo 👍
